modules/auxiliary/admin/vmware/vcenter_forge_saml_token.rb

```diff
@@ -0,0 +1,580 @@
+            'Run',
+            {
+              'Description' => 'Generate vSphere session cookie'
+            }
+          ]
+        ],
+        'DefaultAction' => 'Run',
+        'DefaultOptions' => {
+          'USERNAME' => 'administrator',
+          'DOMAIN' => 'vsphere.local',
+          'RPORT' => 443,
+          'SSL' => true
+        },
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ],
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'SideEffects' => [ IOC_IN_LOGS ]
+        },
+        'Privileged' => true
+      )
+    )
+
+    register_options([
+      OptString.new('RHOSTS', [ true, 'The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit' ]),
+      OptString.new('USERNAME', [ true, 'The username to target using forged credentials', 'administrator' ]),
+      OptString.new('DOMAIN', [true, 'The target vSphere SSO domain', 'vsphere.local']),
+      OptString.new('VC_FQDN', [true, 'DNS FQDN of the vCenter server']),
+      OptPath.new('VC_IDP_CERT', [ true, 'Path to the vCenter IdP certificate' ]),
+      OptPath.new('VC_IDP_KEY', [ true, 'Path to the vCenter IdP private key' ]),
+      OptPath.new('VC_VMCA_CERT', [ true, 'Path to the vCenter VMCA certificate' ])
+    ])
+
+    register_advanced_options([
+      OptPort.new('RPORT', [ true, 'The target port', 443 ]),
+      OptBool.new('SSL', [ false, 'Enable SSL on the connection', true ])
+    ])
+
+    deregister_options('Proxies', 'VHOST')
+  end
+
+  def rhosts
+    datastore['RHOSTS']
+  end
+
+  def rport
+    datastore['RPORT']
+  end
+
+  def username
+    datastore['USERNAME']
+  end
+
+  def domain
+    datastore['DOMAIN']
+  end
+
+  def vcenter_fqdn
+    datastore['VC_FQDN']
+  end
+
+  def vc_idp_cert
+    datastore['VC_IDP_CERT']
+  end
+
+  def vc_idp_key
+    datastore['VC_IDP_KEY']
+  end
+
+  def vc_vmca_cert
+    datastore['VC_VMCA_CERT']
+  end
+
+  def saml_ca_cert
+    datastore['VCENTER_SAML_CA_CERT']
+  end
+
+  def saml_idp_cert
+    datastore['VCENTER_SAML_IDP_CERT']
+  end
+
+  def saml_idp_key
+    datastore['VCENTER_SAML_IDP_KEY']
+  end
+
+  def saml_id
+    datastore['VCENTER_SAML_ID']
+  end
+
+  def saml_issue_instant
+    datastore['VCENTER_SAML_ISSUE']
+  end
+
+  def saml_username
+    datastore['VCENTER_SAML_USER']
+  end
+
+  def saml_domain
+    datastore['VCENTER_SAML_DOMAIN']
+  end
+
+  def saml_response_id
+    datastore['VCENTER_SAML_RESPONSE_ID']
+  end
+
+  def saml_assert_id
+    datastore['VCENTER_SAML_ASSERT_ID']
+  end
+
+  def saml_idx_id
+    datastore['VCENTER_SAML_IDX_ID']
+  end
+
+  def saml_not_before
+    datastore['VCENTER_SAML_NOT_BEFORE']
+  end
+
+  def saml_not_after
+    datastore['VCENTER_SAML_NOT_AFTER']
+  end
+
+  def saml_relay_state
+    datastore['VCENTER_SAML_RELAY_STATE']
+  end
+
+  def sso_token
+    datastore['VCENTER_SAML_TOKEN']
+  end
+
+  def sso_path
+    datastore['VCENTER_SAML_PATH']
+  end
+
+  def ssl
+    datastore['SSL']
+  end
```

Comment:
Using local variables (as per https://github.com/rapid7/metasploit-framework/pull/16484#discussion_r855275898) combined with helper methods may be useful. This is a common code pattern in framework used to fall back to `datastore` options. This is preferred rather than explicitly overriding `datastore` options from within the module.

Here's a similar example from another module:

https://github.com/rapid7/metasploit-framework/blob/140c3bfd50aa0d96d2a823f2ee5e101860c434ea/modules/auxiliary/admin/upnp/soap_portmapping.rb#L57-L83

Reply:
Agreed when needed to offer override that can make sense.

Reply:
Thank you for this explanation. I had seen this pattern in other modules but did not really understand the significance.